### PR TITLE
[IMP] stock_account: Show SN/LN on invoice

### PR DESCRIPTION
**Pull request by aprieels**
In some sectors, the Serial Number needs to be present on invoices.
We add a setting in 'inventory', when it is checked, Add a column to invoice
lines showing the serial number.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/sale_stock/models/account_invoice.py [hunk 1]
```diff
@@ -46,3 +46,11 @@ def _get_anglo_saxon_price_unit(self):
 
     def _compute_average_price(self, qty_done, quantity, moves):
         return self.env['product.product']._compute_average_price(qty_done, quantity, moves)
+
+    def _get_last_step_stock_move_lines(self):
+        """ Overridden from stock_account.
+        Returns the stock moves associated to this invoice line."""
+        rslt = super(AccountInvoiceLine, self)._get_last_step_stock_move_lines()
+        for invoice_line in self:
+            rslt += invoice_line.mapped('sale_line_ids.order_id.picking_ids.move_line_ids')
```
Comment (sle-odoo) [inline addons/sale_stock/models/account_invoice.py]: get the move directly from the sale order line ; same with the puchase line

File: addons/purchase_stock/models/account_invoice.py [hunk 2]
```diff
@@ -148,3 +148,14 @@ def _get_last_step_stock_moves(self):
         for invoice in self.filtered(lambda x: x.type == 'in_refund'):
             rslt += invoice.mapped('invoice_line_ids.purchase_line_id.move_ids').filtered(lambda x: x.state == 'done' and x.location_dest_id.usage == 'supplier')
         return rslt
+
```
Comment (sle-odoo) [inline addons/purchase_stock/models/account_invoice.py]: new line

File: addons/stock_account/security/stock_account_security.xml [hunk 3]
```diff
@@ -14,4 +19,3 @@
     
 </data>
 </odoo>
-
```
Comment (sle-odoo) [inline addons/stock_account/security/stock_account_security.xml]: useless diff

**Comment by sle-odoo** [comment]
- remove invoice_id field
- the method on the po line could directly return the dict {lot_id: qty}
- you should also display the UOM in the report
- this method should get all the stock move line linked to the invoice line **until** the validation date of the invoice (i.e. ignore future changes)
- this should only work in customer invoice or customer refund invoice type
- if customer invoice: count ml in customer ; if customer refund: count ml in stock

**Comment by nhomar** [comment]
Just FYI.

on l10n_mx_edi_** we use this on sales also.

We receive with serials and dispatch with serials in both sides we canneed the serials in the invoice.

It is not a mandatory feature but hardly requested and was lost with the feature **Invoice from picking** on v8 -> v9 transition.

THis feature is nice (i did not review the implementation you did) but see it is on the purchase side only and should be on sales side also.

File: addons/sale_stock/models/account_invoice.py [hunk 4]
```diff
@@ -46,3 +46,13 @@ def _get_anglo_saxon_price_unit(self):
 
     def _compute_average_price(self, qty_done, quantity, moves):
         return self.env['product.product']._compute_average_price(qty_done, quantity, moves)
+
+    def _get_last_step_stock_move_lines(self):
+        """ Overridden from stock_account.
+        Returns the stock moves associated to this invoice line. """
+        rslt = super(AccountInvoiceLine, self)._get_last_step_stock_move_lines()
+        move_lines = self.mapped('sale_line_ids.move_ids.move_line_ids').filtered(lambda ml: ml.state == 'done' and ml.date <= self.create_date)
+        previous_invoices = self.env['account.invoice'].search([('create_date', '<', self.create_date)], order="create_date asc")
```
Comment (sle-odoo) [inline addons/sale_stock/models/account_invoice.py]: mmh?

**Comment by pedrobaeza** [comment]
Hi all, we have been dealing with this for some time as this is indeed a requirement for several sectors, so I want to expose my PoV if it helps: we needed to link specific stock moves to the invoice line instead of trying to deduce them from the sales order to stock move link, as the information otherwise was not correct. You can find in this OCA module (depending on another one - stock_picking_invoice_link -), the result: https://github.com/OCA/account-invoice-reporting/tree/11.0/account_invoice_production_lot.

There's still pending the returns management however.

About your PR, are you taking into account that this should be only shown for customer invoices? What happens with refunds?

**Comment by sle-odoo** [comment]
closing in favour of https://github.com/odoo/odoo/pull/33022

**Pull request closed without merging**